source/Date.cpp

```diff
@@ -49,23 +52,57 @@ Date::Date(int day, int month, int year)
 // Convert a date to a string.
 const string &Date::ToString() const
 {
+	string dateFormat = Preferences::DateFormat();
+
+	if(dateFormat != dateFormatInUse)
+	{
+		dateFormatInUse = dateFormat;
+		str = "";
+	}
+
 	// Because this is a somewhat "costly" operation, cache the result. The
 	// cached value is discarded if the date is changed.
-	if(date && str.empty())
+	if((date && str.empty()))
 	{
 		int day = Day();
 		int month = Month();
 		int year = Year();
 
-		str = Weekday(day, month, year);
-		str.append(", ");
-		str.append(to_string(day));
-		str.append(" ");
 		static const string MONTH[] = {
-			"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
-		str.append(MONTH[month - 1]);
-		str.append(" ");
-		str.append(to_string(year));
+				"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
+
+		if(dateFormat == "ymd")
+		{
+			str.append(to_string(year));
+			str.append("-");
+			if (to_string(month).length() == 1)
+				str.append("0");
+			str.append(to_string(month));
+			str.append("-");
+			if (to_string(day).length() == 1)
+				str.append("0");
+			str.append(to_string(day));
+		}
+		else if(dateFormat == "mdy")
+		{
+			str = Weekday(day, month, year);
+			str.append(" ");
+			str.append(MONTH[month - 1]);
+			str.append(" ");
+			str.append(to_string(day));
+			str.append(", ");
+			str.append(to_string(year));
+		}
+		else if(dateFormat == "dmy")
+		{
+			str = Weekday(day, month, year);
+			str.append(", ");
+			str.append(to_string(day));
+			str.append(" ");
+			str.append(MONTH[month - 1]);
+			str.append(" ");
+			str.append(to_string(year));
+		}
```

Comment:
Why don't you convert the numbers into the components of the final string first using some temporary variables in this method and then only do the combination of those strings into the appropriate order based on the preference in if statements (though, once you use an enum class for the preference, you should probably use a switch-case statement.)

Reply:
I've rewritten that code now.